### PR TITLE
[FLINK-30368][table-planner] Fix calcite method RelMdUtil$numDistinctVals() wrongly return zero if the method input domainSize much larger than numSelected

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change
This pr is aims to fix calcite method `RelMdUtil$numDistinctVals()` wrongly return zero if the method input `domainSize` is a very large value.  If `domainSize` is a very large value (like 2 x 10^16), Math.log(1.0 - 1.0 / dSize) will return 0, and `domainSize` will be set to 0, which is unreasonable in this scenario. (This bug will let join type choose `NestedLoopJoin` instead of `HashJoin` in physical program).

## Brief change log
- Class `RelMdUtil` is copied from calcite `org.apache.calcite.rel.metadata.RelMdUtil`, and modified method `numDistinctVals()`.

## Verifying this change

No tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented?  no

### CODE REVIEW

**Comment by snuyanzin** [comment]
Thanks for your contribution
Could you please clarify how we can be sure that this change does what expected?
May be some tests could be helpful to have.

Also it would be helpful to have info about what lines are changed in this class, like it is done e.g. in `RexSimplify`, `SqlToRelConverter` and other Calcite's classes

And another question: shouldn't this issue be raised in Calcite's jira?

**Comment by swuferhong** [comment]
> Thanks for your contribution Could you please clarify how we can be sure that this change does what expected? May be some tests could be helpful to have.
> 
> Also it would be helpful to have info about what lines are changed in this class, like it is done e.g. in `RexSimplify`, `SqlToRelConverter` and other Calcite's classes
> 
> And another question: shouldn't this issue be raised in Calcite's jira?

Hi, This bug was found in the `tpc-ds` benchmark tests. So it was difficult to reproduce in Flink ut test case, but I will try to reproduce it .

OK, I will add more information in class annotation.

I think it necessary to open a jira in calcite, and I will do it, Thanks!

**Comment by swuferhong** [comment]
Caicite jira link: https://issues.apache.org/jira/browse/CALCITE-5431

**Comment by godfreyhe** [comment]
@swuferhong Currently, we had copied `RelMdUtil#numDistinctVals` to `FlinkRelMdUtil.numDistinctVals` and do some changes. does it not meet your requirements ？

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala [hunk 1]
```diff
@@ -247,6 +248,77 @@ object FlinkRelMdUtil {
...
+   *   number of distinct rows
+   */
+  def getJoinDistinctRowCount(
+      mq: RelMetadataQuery,
+      joinRel: RelNode,
+      joinType: JoinRelType,
+      groupKey: ImmutableBitSet,
+      predicate: RexNode,
+      useMaxNdv: Boolean): Double = {
+    if ((predicate == null || predicate.isAlwaysTrue) && groupKey.isEmpty) return 1d
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala:279]: please format the copied code

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala [hunk 2]
```diff
@@ -247,6 +248,77 @@ object FlinkRelMdUtil {
     RexUtil.composeConjunction(rexBuilder, pushable, true)
   }
 
+  /**
+   * This method is copied from calcite RelMdUtil. This method should be removed once CALCITE-4351
+   * is fixed. See CALCITE-4351 and FLINK-19780.
+   *
+   * Computes the number of distinct rows for a set of keys returned from a join. Also known as NDV
+   * (number of distinct values).
+   *
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala:257]: please also mark which lines are changed

**Review by godfreyhe — COMMENTED** [review]
Is there any plan affected?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala (hunk 1 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala:279]: > please format the copied code

Done!

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala (hunk 2 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRelMdUtil.scala:257]: > please also mark which lines are changed

Done!

**Review by godfreyhe — APPROVED** [review]
LGTM

**Pull request closed without merging**